File: src/my_cpp_pkg/src/battery.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/srv/set_led.hpp"

using namespace rclcpp;
using namespace std;

class BatteryNode : public Node 
{
public:
    BatteryNode() : Node("battery"), battery_state(false)
    {
        //get init time
        last_time_battery_state_changed_ = this->get_clock()->now().seconds();
        timer_ = this->create_wall_timer(chrono::milliseconds(100),bind(&BatteryNode::check_led_state,this));
        RCLCPP_INFO(this->get_logger(),"Battery has been started.");
    }

private:
    void set_led(int64_t led_number, bool state)
    {
        threads_.push_back(thread(bind(&BatteryNode::call_battery_client,this,led_number,state)));
    }

    void call_battery_client(int64_t led_number, bool state)
    {
        battery_client_ = this->create_client<my_robot_interfaces::srv::SetLed>("set_led");
        while (!battery_client_->wait_for_service(chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(),"Waiting for set led service...");
        }
        //init request
        auto request = std::make_shared<my_robot_interfaces::srv::SetLed::Request>();
        request->led_number = led_number;
        request->state = state;

        auto future = battery_client_->async_send_request(request);
        try
        {
            auto response = future.get();       //this will block the thread
            RCLCPP_INFO(this->get_logger(),"%d",response->success);
        }
        catch(const std::exception& e)
        {
            RCLCPP_ERROR(this->get_logger(),"Service call failed.");
        }
    }

    void check_led_state()
    {
        double clock = this->get_clock()->now().seconds();
        if(battery_state)   // battery is full
        {
            if(clock - last_time_battery_state_changed_ > 4.0){
                RCLCPP_INFO(this->get_logger(),"Battery is empty! Charging now...");
                last_time_battery_state_changed_ = clock;
                battery_state = false;   // change battery full to empty state
                set_led(3,true);
            }
        }
        else        // battery is empty
        {
            if(clock - last_time_battery_state_changed_ > 6.0){
                RCLCPP_INFO(this->get_logger(),"Battery is full! Stop charging now...");
                last_time_battery_state_changed_ = clock;
                battery_state = true;   // change battery empty to full state
                set_led(3,false);
            }
        }
        
    }

    Client<my_robot_interfaces::srv::SetLed>::SharedPtr battery_client_;
    TimerBase::SharedPtr timer_;
    vector<thread> threads_;
    bool battery_state;     //true: fully charge, false: empty
    double last_time_battery_state_changed_;
};

int main(int argc, char **argv)
{
    init(argc, argv);
    auto node = std::make_shared<BatteryNode>(); 
    spin(node);
    shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/battery.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from my_robot_interfaces.srv import SetLed

class BatteryNode(Node):
    def __init__(self):
        super().__init__("battery") 
        self.battery_state = "full"
        #get init time
        self.last_time_battery_state_changed_ = self.get_current_time()
        self.timer_ = self.create_timer(0.1,self.check_led_state)
        self.get_logger().info("Battery has been started.")
    
    def get_current_time(self):
        secs,nanosecs = self.get_clock().now().seconds_nanoseconds()
        return secs + nanosecs/1000000000.0
    
    def check_led_state(self):
        clock = self.get_current_time()
        if self.battery_state == "full":
            if clock - self.last_time_battery_state_changed_ > 4.0 :
                self.get_logger().info("Battery is empty! Charging now...")
                self.last_time_battery_state_changed_ = clock
                self.battery_state = "empty"    #change battery full to empty state
                self.call_battery_client(3,True)
        else:
            if clock - self.last_time_battery_state_changed_ > 6.0:
                self.get_logger().info("Battery is full! Stop charging now...")
                self.last_time_battery_state_changed_ = clock
                self.battery_state = "full"     #change battery empty to full state
                self.call_battery_client(3,False)

    def call_battery_client(self,led_number,state):
        client_ = self.create_client(SetLed,"set_led")
        while not client_.wait_for_service(1) :
            self.get_logger().warn("Waiting for set led service...")
        request = SetLed.Request()          #init request from service type
        request.led_number = led_number
        request.state = state

        future = client_.call_async(request)
        # add callback_add_two_ints to be executed when client.call_async(request) is done
        future.add_done_callback(self.callback_battery_client)   #nho gan gia tri

    def callback_battery_client(self,future):
        try:
            response = future.result()
            self.get_logger().info(str(response.success))
        except Exception as e:
            self.get_logger().error("Service call failed %r" %(e,))


def main(args=None):
    rclpy.init(args=args)
    node = BatteryNode() 
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()